### PR TITLE
Remove .la files generated by autotools

### PR DESCRIPTION
**Pull request by robert-ancell**
If you build a part that depends on a library part, e.g.

parts:
    glib:
        plugin: autotools
        source: https://download.gnome.org/core/3.20/3.20.2/sources/glib-2.48.1.tar.xz
    glib-networking:
        plugin: autotools
        source: https://download.gnome.org/core/3.20/3.20.2/sources/glib-networking-2.48.2.tar.xz
        after:
          - glib

Then the dependant part fails to link due to the .la files being moved around:

  CCLD     glib-pacrunner
libtool: warning: library '/home/bob/bzr/gnome-3-20-snap/stage/lib/libgio-2.0.la' was moved.
libtool:   error: cannot find the library '/lib/libgobject-2.0.la' or unhandled argument '/lib/libgobject-2.0.la'

.la files aren't commonly used (they are removed in Debian) and removing them fixes this issue. See discussion http://stackoverflow.com/questions/7104494/building-so-module-with-autotools-libtool-without-la-and-a-variants-being-ins

### CODE REVIEW

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by robert-ancell** [comment]
I'm not sure if we should allow the user to keep the .la files or not. The current patch makes it impossible to keep them. I think they should be removed by default though as this is what most people will want and it avoids the original issue with building against them.

**Comment by robert-ancell** [comment]
$ cat ./stage/lib/libgio-2.0.la
# libgio-2.0.la - a libtool library file
# Generated by libtool (GNU libtool) 2.4.6 Debian-2.4.6-0.1
#
# Please DO NOT delete this file!
# It is necessary for linking the library.

# The name that we can dlopen(3).
dlname='libgio-2.0.so.0'

# Names of this library.
library_names='libgio-2.0.so.0.4800.1 libgio-2.0.so.0 libgio-2.0.so'

# The name of the static archive.
old_library=''

# Linker flags that cannot go in dependency_libs.
inherited_linker_flags=''

# Libraries that this one depends upon.
dependency_libs=' /lib/libgobject-2.0.la -lffi /lib/libgmodule-2.0.la -ldl /lib/libglib-2.0.la -lpcre -lpthread -lz -lselinux -lresolv'

# Names of additional weak libraries provided by this library
weak_library_names=''

# Version information for libgio-2.0.
current=4800
age=4800
revision=1

# Is this an already installed library?
installed=yes

# Should we warn about portability when linking against -modules?
shouldnotlink=no

# Files to dlopen/dlpreopen
dlopen=''
dlpreopen=''

# Directory that this library needs to be installed in:
libdir='/lib'

**Comment by robert-ancell** [comment]
``$ cat ./stage/lib/libgio-2.0.la
# libgio-2.0.la - a libtool library file
# Generated by libtool (GNU libtool) 2.4.6 Debian-2.4.6-0.1
#
# Please DO NOT delete this file!
# It is necessary for linking the library.

# The name that we can dlopen(3).
dlname='libgio-2.0.so.0'

# Names of this library.
library_names='libgio-2.0.so.0.4800.1 libgio-2.0.so.0 libgio-2.0.so'

# The name of the static archive.
old_library=''

# Linker flags that cannot go in dependency_libs.
inherited_linker_flags=''

# Libraries that this one depends upon.
dependency_libs=' /lib/libgobject-2.0.la -lffi /lib/libgmodule-2.0.la -ldl /lib/libglib-2.0.la -lpcre -lpthread -lz -lselinux -lresolv'

# Names of additional weak libraries provided by this library
weak_library_names=''

# Version information for libgio-2.0.
current=4800
age=4800
revision=1

# Is this an already installed library?
installed=yes

# Should we warn about portability when linking against -modules?
shouldnotlink=no

# Files to dlopen/dlpreopen
dlopen=''
dlpreopen=''

# Directory that this library needs to be installed in:
libdir='/lib'`

**Comment by robert-ancell** [comment]
```
$ cat ./stage/lib/libgio-2.0.la
# libgio-2.0.la - a libtool library file
# Generated by libtool (GNU libtool) 2.4.6 Debian-2.4.6-0.1
#
# Please DO NOT delete this file!
# It is necessary for linking the library.

# The name that we can dlopen(3).
dlname='libgio-2.0.so.0'

# Names of this library.
library_names='libgio-2.0.so.0.4800.1 libgio-2.0.so.0 libgio-2.0.so'

# The name of the static archive.
old_library=''

# Linker flags that cannot go in dependency_libs.
inherited_linker_flags=''

# Libraries that this one depends upon.
dependency_libs=' /lib/libgobject-2.0.la -lffi /lib/libgmodule-2.0.la -ldl /lib/libglib-2.0.la -lpcre -lpthread -lz -lselinux -lresolv'

# Names of additional weak libraries provided by this library
weak_library_names=''

# Version information for libgio-2.0.
current=4800
age=4800
revision=1

# Is this an already installed library?
installed=yes

# Should we warn about portability when linking against -modules?
shouldnotlink=no

# Files to dlopen/dlpreopen
dlopen=''
dlpreopen=''

# Directory that this library needs to be installed in:
libdir='/lib'`

```

**Comment by robert-ancell** [comment]
The above file shows a generated .la file with the absolute /lib paths that don't work.

**Comment by robert-ancell** [comment]
Didier pointed out this can be achieved using:
parts:
  stage: [ -*.la ]
  snap: [ -*.la ]

However this PR is probably still worth considering as this is tedious to add everywhere and it's unlikely the .la files will be useful / work anyway.

**Comment by robert-ancell** [comment]
Actually, now I'm not sure that the stage filter is working..

File: snapcraft/plugins/autotools.py [hunk 1]
```diff
@@ -117,3 +118,6 @@ def build(self):
         self.run(configure_command + self.options.configflags)
         self.run(['make', '-j{}'.format(self.project.parallel_build_count)])
         self.run(make_install_command)
+
+        # Remove .la files which don't work when they are moved around
+        subprocess.check_call("find . -name '*.la' | xargs rm", shell=True, cwd=self.installdir)
```
Comment (sergiusens) [inline snapcraft/plugins/autotools.py]: can you change this to use `os.walk` please?

**Comment by sergiusens** [comment]
ok to test

**Comment by sergiusens** [comment]
@robert-ancell so bad luck:

```
snapcraft/plugins/autotools.py:92:1: C901 'AutotoolsPlugin.build' is too complex (11)
```

**Comment by sergiusens** [comment]
maybe just separate the new code into a helper method of the class or private method of the module or in `snapcraft.internal.common`

`AutotoolsPlugin._remove_la()`
`autotools._remove_la(directory)`
`snapcraft.internal.common.remove_files(base_directory, patterns)`

**Comment by sergiusens** [comment]
snapcraft/plugins/autotools.py:128:30: E211 whitespace before '('

**Comment by sergiusens** [comment]
retest this please

**Pull request merged**